packages/internal/src/upgrade-api.js

```diff
@@ -0,0 +1,41 @@
+// @ts-check
+// @jessie-check
+import { isObject } from '@endo/marshal';
+
+/**
+ * @typedef {{ name: string, upgradeMessage: string, incarnationNumber: number }} DisconnectionObject
+ */
+
+/**
+ * Makes an Error-like object for use as the rejection value of promises
+ * abandoned by upgrade.
+ *
+ * @param {string} upgradeMessage
+ * @param {number} toIncarnationNumber
+ * @returns {DisconnectionObject}
+ */
+export const makeUpgradeDisconnection = (upgradeMessage, toIncarnationNumber) =>
+  harden({
+    name: 'vatUpgraded',
+    upgradeMessage,
+    incarnationNumber: toIncarnationNumber,
+  });
+harden(makeUpgradeDisconnection);
+
+// TODO: Simplify once we have @endo/patterns (or just export the shape).
```

Comment:
We do have `@endo/patterns` now.

Reply:
My first attempt didn't quite work, so I'm deferring this until we've got a little more time to spend on it.